Joss paper (#530)
CEED/libCEED: doc/papers/joss/paper.md

    affiliation: 1
  - name: XXXX TODO Many more authors to be added XXXX
    affiliation: 2
affiliations:
 - name: University of Colorado Boulder
   index: 1
 - name: Lawrence Livermore National Laboratory
   index: 2
date: XXXX TODO write date XXXX
bibliography: paper.bib

# Optional fields if submitting to a AAS journal too, see this blog post:
# https://blog.joss.theoj.org/2018/12/a-new-collaboration-with-aas-publishing
aas-doi: 10.3847/xxxxx <- update this with the DOI from AAS once you know it.
aas-journal: Astrophysical Journal <- The name of the AAS journal.
---

# Summary

High-order numerical methods are widely used in Partial Differential Equation (PDE)
solvers, but software packages that provide high-performance implementations have
often been special-purpose and intrusive. ``libCEED`` is an open-source light-weight
library that offers a purely algebraic interface for efficient operator evaluation
and matrix-free preconditioning ingredients. libCEED supports run-time selection of
implementations tuned for a variety of computational architectures, including CPUs and
GPUs, and can be unobtrusively integrated in new and legacy software to provide portable

> performance. We introduce ``libCEED``’s conceptual framework and interface, and show examples of its integration with other packages, including PETSc and MFEM.

Reviewer: Do we want to mention Nek as well? We aren't highly integrated, but we have examples and its a well known code.

Author: Sure, please feel free to make all edits and give contribution as you like. I'm not going to be the only one authoring this

Reviewer: I wanted to discuss this one, since we don't have tight Nek integration like we do with MFEM and PETSc. I could do either way on including, leaning towards yes?

Author: I would say, yes. And we don't need to be specific in this sentence, since it says "including". Or it could be something like:  
```suggestion
performance. We introduce ``libCEED``’s conceptual framework and interface, and show examples of its integration with other packages, such as PETSc, MFEM, and Nek5000.
```

Reviewer: I like that.

In finite element formulations, the weak form of a PDE is evaluated on a subdomain
(element), and the local results are composed into a larger system of equations that
models the entire problem. In particular, when high-order finite elements or spectral
elements are used, the resulting sparse matrix representation of the global operator
is computationally expensive, with respect to both the memory transfer and floating
point operations needed for its evaluation [@Brown:2010]. ``libCEED`` provides an
interface for matrix-free operator description that enables efficient evaluation on
a variety of computational device types (selectable at run time).

# libCEED's API

> ``libCEED``'s Application Programming Interface (API) provides the local action of the

Reviewer: Do we want to mention C99/F77/Python interface. Not sure if we want to direct people to the binder?

Author: Yeah, this was one of my questions on Zulip. It is up to us to decide. I would totally mention the interfaces. To what extent to describe them is totally up to us. If we want to describe the Python one a bit better, I have some paragraphs ready that I wrote for the SciPy submission

Author: Done in [d6f5a2d](https://github.com/CEED/libCEED/pull/530/commits/d6f5a2d9c1e6f05447ee70f32025e58102c9e145)

Fig. \ref{fig:libCEEDBackends} shows a subset of the backend implementations available
in ``libCEED``. GPU implementations are available via pure CUDA as well as the OCCA
and MAGMA libraries. CPU implementations are available via pure C and AVX intrinsics
as well as the LIBXSMM library. libCEED provides a unified interface, so that users
only need to write a single source code and can select the desired specialized
implementation at run time. Moreover, each process or thread can instantiate an
arbitrary number of backends.

# Performance Benchmarks

The Center for Efficient Exascale Discretizations (CEED), part of the Exascale Computing Project (ECP) uses Benchmark Problems (BPs) to test and compare the performance of high-order finite element implementations. In Fig. \ref{fig:Skylakeselfxsmm}, we analyze the performance of libCEED's LIBXSMM backend on a Poisson's problem. We measure performance over 20 iterations of unpreconditioned Conjugate Gradient (CG) and plot throughput, versus execution time per iteration, for different values of the polynomial degree $p$.

![\texttt{xsmm/serial} backend on a Skylake

> (2x Intel Xeon Platinum 8180M CPU 2.50GHz).\label{fig:Skylakeselfxsmm}](img/plot_libCEED_PETScBP3_cpuselfxsmmserial_N001_pn56_time.pdf)

Reviewer: I think the /xsmm/* and /cuda/gen bechnmarks are good here. BP3?

Author: It was just an old figure from another 2-page summary produced for a conference submission. Totally up to us to decide which and how many results to show (which was one of the checkboxes above)

Reviewer: Cool. I'm just typing out thoughts as they occur to me, but you were totally ahead of me on this one.

Reviewer: I think that we have enough small changes in performance that we should do fresh runs. I'm thinking BP3 or BP1 + BP3 in the following configurations:

1) `/gpu/cuda/gen` + V100
2) `/cpu/self/xsmm/serial` + Skylake Platinum
3) `/cpu/self/xsmm/blocked` + Skylake Platinum

The benchmark script uses the `bpsraw` right now, but I'm honestly not sure if `bpsraw` or `bps` runs faster.

Author: By a very quick check on my machine, without even running the whole battery for all degrees like in the benchmark script, but only one configuration, the `bpsraw` is actually slower than the `bps`

Reviewer: I think I used `bpsraw` at first because `bps` originally had a pretty big setup time for the mesh. Since its faster, I think we can switch over (though, we'd have to update the script to run based on number of box faces or compute faces out of local dofs desired.

Author: I can't upload the log files for some reason now in this thread.. But yes, for my very quick test on my machine, from PETSc' l`-og_view` summary (the Max column) I've got:
bpsraw in serial:
Time (sec):           5.403e-02
bps in serial:
Time (sec):           2.709e-02

bpsraw in parallel (with `mpiexec -n 2`):
Time (sec):           3.220e-02
bps in parallel (with `mpiexec -n 2`):
Time (sec):           1.685e-02

Author: They were just run as `./bpsraw -problem bp3` and `./bps -problem bp3`, and ` mpiexec -n 2 ./bpsraw -problem bp3` and ` mpiexec -n 2 ./bps -problem bp3`, respectively, without any other specified option

Reviewer: I still think that `bps` is faster, but the default sizes don't match.
Closer lvec sized test: (fiddling with `-local` to get `-cells` is a pain)
```
[jeremy@sinensis petsc]$ ./bpsraw -problem bp3 -degree 2 -local 9261

-- CEED Benchmark Problem 3 -- libCEED + PETSc --
  PETSc:
    PETSc Vec Type                     : seq
  libCEED:
    libCEED Backend                    : /cpu/self/xsmm/blocked
    libCEED Backend MemType            : host
    libCEED User Requested MemType     : none
  Mesh:
    Number of 1D Basis Nodes (p)       : 3
    Number of 1D Quadrature Points (q) : 4
    Global nodes                       : 10773
    Process Decomposition              : 1 1 1
    Local Elements                     : 1170 = 9 10 13
    Owned nodes                        : 10773 = 19 21 27
    DoF per node                       : 1
  KSP:
    KSP Type                           : cg
    KSP Convergence                    : CONVERGED_RTOL
    Total KSP Iterations               : 9
    Final rnorm                        : 2.149739e-12
  Performance:
    Pointwise Error (max)              : 2.327021e-03
    CG Solve Time                      : 0.00627679 (0.00627679) sec
[jeremy@sinensis petsc]$ ./bps -problem bp3 -cells 10,10,10 -degree 2

-- CEED Benchmark Problem 3 -- libCEED + PETSc --
  PETSc:
    PETSc Vec Type                     : seq
  libCEED:
    libCEED Backend                    : /cpu/self/xsmm/blocked
    libCEED Backend MemType            : host
    libCEED User Requested MemType     : none
  Mesh:
    Number of 1D Basis Nodes (p)       : 3
    Number of 1D Quadrature Points (q) : 4
    Global nodes                       : 6859
    Owned nodes                        : 6859
    DoF per node                       : 1
  KSP:
    KSP Type                           : cg
    KSP Convergence                    : CONVERGED_RTOL
    Total KSP Iterations               : 8
    Final rnorm                        : 1.379813e-12
  Performance:
    Pointwise Error (max)              : 5.054563e-03
    CG Solve Time                      : 0.00467284 (0.00467284) sec
```

Author: Besides the solve time... It's also interesting the double-sized error, isn't it? 🤔

Reviewer: Maybe because of the extra CG iteration the bpsraw error goes down slightly. As we increase mesh size I'd imagine the difference goes down.

Author: Yeah maybe... Up to us whether we want to modify the benchmark script to run `bps` instead of `bpsraw` and compare results for all problem sizes to make sure that `bps` actually perform better

Reviewer 2: It makes sense for `bps` (after initialization) to be slightly faster because of how it handles boundary conditions.

Author: > I think that we have enough small changes in performance that we should do fresh runs. I'm thinking BP3 or BP1 + BP3 in the following configurations:
> 
> 1. `/gpu/cuda/gen` + V100
> 2. `/cpu/self/xsmm/serial` + Skylake Platinum
> 3. `/cpu/self/xsmm/blocked` + Skylake Platinum
> 
Hey, @jeremylt @jedbrown I have run the benchmarks using the `petsc-bps` on CU Summit's skylake partition (the only Skylake node I have access to). The [results](https://drive.google.com/open?id=1GCgFZG4_Oi6Ad3eF6dHBJ11edwGbBh5_) are not better than what we had already that @jedbrown tested last summer (I don't know on which machine).

Please let me know what other tests I could run, as I was hoping we could select some of the updated results also for the SciPy submission, which is coming up. Thanks.

Reviewer: I think CU has Skylake Golds, so they won't be as good as @jedbrown's runs on Skylake Platinums.

Reviewer: Yeah, their documentation on hardware is spotty.

Here is what I've found:
https://www.colorado.edu/rc/news/past/summitcondo

Their Summit specs are out of date:
https://www.colorado.edu/rc/resources/summit/specifications

Reviewer 3: You can always use `cat /proc/cpuinfo`. 😁

# Performance Benchmarks

The Center for Efficient Exascale Discretizations (CEED), part of the Exascale Computing Project (ECP) uses Benchmark Problems (BPs) to test and compare the performance of high-order finite element implementations. In Fig. \ref{fig:Skylakeselfxsmm}, we analyze the performance of libCEED's LIBXSMM backend on a Poisson's problem. We measure performance over 20 iterations of unpreconditioned Conjugate Gradient (CG) and plot throughput, versus execution time per iteration, for different values of the polynomial degree $p$.

![\texttt{xsmm/serial} backend on a Skylake
(2x Intel Xeon Platinum 8180M CPU 2.50GHz).\label{fig:Skylakeselfxsmm}](img/plot_libCEED_PETScBP3_cpuselfxsmmserial_N001_pn56_time.pdf)

# Applications

> To highlight the ease of library reuse for solver composition and leverage libCEED's full capability for real-world applications, libCEED comes with a suite of application examples, including problems of interst to the fluid dynamics and continuum mechanics communities.

Reviewer 4: ```suggestion
To highlight the ease of library reuse for solver composition and leverage libCEED's full capability for real-world applications, libCEED comes with a suite of application examples, including problems of interest to the fluid dynamics and continuum mechanics communities.
```

# Acknowledgements

We acknowledge the US Department of Energy (DOE) Exascale Computing Project (ECP)
(17-SC-20-SC) and thank all collaborators.

> 

Reviewer 4: The full ECP disclaimer is:

> ECP Disclaimer: This research is supported by the Exascale Computing Project (17-SC-20-SC), a collaborative effort of two U.S. Department of Energy organizations (Office of Science and the National Nuclear Security Administration) responsible for the planning and preparation of a capable exascale ecosystem, including software, applications, hardware, advanced system engineering and early testbed platforms, in support of the nations exascale computing imperative.

Author: Thanks. Updated.

localizes DOFs based on the elements, $B$ defines the action of the basis functions
(or their gradients) on the nodes, and $D$ is the user-defined pointwise function
describing the physics of the problem at the quadrature points, also called the
QFunction. QFunctions, which can either be defined by the user or selected from a
gallery of available built-in functions in the library, are pointwise functions
that do not depend on element resolution, topology, or basis degree (selectable
at run time). This easily allows $hp$-refinement studies (where $h$ commonly denotes the average element size and $p$ the polynomial degree of the basis functions in 1D) and $p$-multigrid solvers. libCEED also supports composition of different operators for multiphysics problems and mixed-element meshes (see Fig. \ref{fig:schematic}).

![A schematic of element restriction and basis applicator operators for
elements with different topology. This sketch shows the independence of QFunctions
(in this case representing a Laplacian) element resolution, topology, or basis degree.\label{fig:schematic}](img/QFunctionSketch.pdf)

> LibCEED is a C99 library with Fortran77 and Python interfaces. The Python interface was developed using the C Foreign Function Interface (CFFI) for Python. CFFI allows to reuse most of the C declarations and requires only a minimal adaptation of some of them. The C and Python APIs are mapped in a nearly 1:1 correspondence. For instance, a ``CeedVector`` object is exposed as ``libceed.Vector`` in Python, and may reference memory that is also accessed via Python arrays from the NumPy [@NumPy] or Numba [@Numba] packages, for handling host or device memory (when interested in GPU computations with CUDA). Flexible pointer handling in libCEED makes it easy to provide zero-copy host and (GPU) device support for any desired Python array container.

Reviewer: Do we want to reference our SciPy paper here?

Author: I can do it as soon as I have time later. Do you want to add the latest developments on the scaling for multigrid methods, and QFunction context? (I think a part from the possible file organization, the meat in PR #596 will not change much, right?)

Author: Done and rebased onto `main`.

Author: @jeremylt  do you want to add those two things in here? 
I think we forgot to mention the changes in #596 and #579  in the release notes page

Reviewer: Since we don't talk about vectors, except briefly for the Python interface, I don't know that the context object merits much discussion.

To achieve high performance, libCEED can take advantage of a tensor-product
finite-element basis and quadrature rule to apply the action of the basis
operator $B$ or, alternatively, efficiently operate on bases that are defined
on arbitrary-topology elements. Furthermore, the algebraic decomposition described in
Eq. (\ref{eq:decomposition}) can represent either linear/nonlinear or
symmetric/asymmetric operators and exposes opportunities for device-specific
optimizations.

![libCEED is a low-level API for finite element codes, that has specialized implementations
(backends) for heterogeneous architectures.\label{fig:libCEEDBackends}](img/libCEEDBackends.pdf)

> Fig. \ref{fig:libCEEDBackends} shows a subset of the backend implementations (backends) available in libCEED and its role, as a low-level library that allows a wide variety of applications to share highly optimized discretization kernels.

Reviewer: Will the figure need an update?

Author: Updated with your latest version. Thanks.

> GPU implementations are available via pure CUDA [@CUDAwebsite] as well as the OCCA [@OCCAwebsite] and MAGMA [@MAGMAwebsite] libraries. CPU implementations are available via pure C and AVX intrinsics as well as the LIBXSMM library [@LIBXSMM]. libCEED provides a unified interface, so that users only need to write a single source code and can select the desired specialized implementation at run time. Moreover, each process or thread can instantiate an arbitrary number of backends.

Reviewer: ```suggestion
GPU implementations are available via pure CUDA [@CUDAwebsite] and pure HIP [@HIPwebsite] as well as the OCCA [@OCCAwebsite] and MAGMA [@MAGMAwebsite] libraries. CPU implementations are available via pure C and AVX intrinsics as well as the LIBXSMM library [@LIBXSMM]. libCEED provides a unified interface, so that users only need to write a single source code and can select the desired specialized implementation at run time. Moreover, each process or thread can instantiate an arbitrary number of backends.
```

Author: Hey Jeremy, I'm off this week. Please make any change you think appropriate.

`libCEED` provides portable performance via run-time selection of implementations optimized for CPUs and GPUs, including support for just-in-time (JIT) compilation.
It is designed for convenient use in new and legacy software, and offers interfaces in C99 [@C99-lang], Fortran77 [@Fortran77-lang], Python [@Python-lang], Julia [@Julia-lang], and Rust [@Rust-lang].
Users and library developers can integrate `libCEED` at a low level into existing applications in place of existing matrix-vector products without significant refactoring of their own discretization infrastructure.
Alternatively, users can utilize integrated `libCEED` support in MFEM [@MFEMlibrary; @mfem-paper].

In addition to supporting applications and discretization libraries, `libCEED` provides a platform for performance engineering and co-design, as well as an algebraic interface for solvers research like adaptive $p$-multigrid, much like how sparse matrix libraries enable development and deployment of algebraic multigrid solvers.

# Concepts and interface

> Consider finite element discretization of a problem based on a weak form with one weak derivative: find $u$ such that

Author: ```suggestion
Consider the finite element discretization of a problem based on a weak form with one weak derivative: find $u$ such that
```

or 
```suggestion
Consider a finite element discretization of a problem based on a weak form with one weak derivative: find $u$ such that
```

?

Reviewer 2: I'm okay with my form or either of yours. I think my wording is more process oriented, analogous to solving a problem "by factorization" instead of "a factorization" or "the factorization". [This page](https://en.wikipedia.org/wiki/LU_decomposition) uses both forms of "factorization" liberally, for example.

Author: I'm not the native English speaker here :) I was reading it more like a regular noun, so an article sounded appropriate to me, but really, I entrust you

Additionally, a single source implementation (in vanilla C or C++) for the `CeedQFunction`s can be used on CPUs or GPUs (transparently using the @NVRTCwebsite, HIPRTC, or OCCA [@OCCAwebsite] run-time compilation features).

`libCEED` provides computation of the true operator diagonal for preconditioning with Jacobi and Chebyshev as well as direct assembly of sparse matrices (e.g., for coarse operators in multigrid) and construction of $p$-multigrid prolongation and restriction operators.
Preconditioning matrix-free operators is an active area of research; support for domain decomposition methods and inexact subdomain solvers based on the fast diagonalization method [@lottes2005hms] are in active development.

![A schematic of element restriction and basis applicator operators for
elements with different topology. This sketch shows the independence of Q-functions

> (in this case representing a Laplacian) on element resolution, topology, and basis degree.\label{fig:schematic}](img/QFunctionSketch.pdf)

Author: ```suggestion
(in this case representing a Laplacian operator) on element resolution, topology, and basis degree.\label{fig:schematic}](img/QFunctionSketch.pdf)
```
?

Reviewer: Personally, I've heard `the Laplacian` or the `Laplace operator`

Author: Sounds good to leave it as is to me, too. I was just proof reading, double checking things and wanted to check with others if it sounded ok

Reviewer 2: No preference.

  - spectral elements
  - matrix-free
authors:
  - name: Valeria Barra
    orcid: 0000-0003-1129-2056
    affiliation: 1
  - name: Natalie Beams
    affiliation: 3
  - name: Jed Brown
    orcid: 0000-0002-9945-0639
    affiliation: 1
  - name: Jean-Sylvain Camier
    affiliation: 2
  - name: Veselin Dobrev
    affiliation: 2
  - name: Yohann Dudouit
    affiliation: 2
  - name:  Leila Ghaffari
    affiliation: 1
  - name: Tzanio Kolev
    affiliation: 2
  - name: David Medina
    affiliation: 4
  - name: Thilina Rathnayake
    affiliation: 5
  - name: Jeremy Thompson
    affiliation: 1
  - name: Stan Tomov

>     affiliation: 3

Reviewer 2: @nbeams @camierjs @v-dobrev @YohannDudouit @LeilaGhaffari @tzanio @dmed256 @thilinarmtb @jeremylt @stomov We've tentatively added you to this author list because you are listed as an author on the users manual. Could you please contribute your ORCID (please create an account at https://orcid.org if you don't have one)? (You can leave a comment here and we'll add it, or you can push it in a commit.) By doing so, you're confirming your willingness to be an author pursuant to [JOSS authorship guidelines](https://joss.readthedocs.io/en/latest/submitting.html#authorship). Also, please let us know if anyone has been missed.

Reviewer 2: @abdelfattah83 Cc: #545

Reviewer: @pazner Since the plan is to include the Julia interface in the JOSS submission, can you add your ORCID here?

  - spectral elements
  - matrix-free
authors:

>   - name: Ahmad Abdelfattah

Author: Hi @abdelfattah83 thanks for adding yourself and welcome! Yes, you assumed right. Our first idea was to include all authors from the user-manual (who, in turn, are all authors who satisfy both points in our [CONTRIBUTING](https://github.com/CEED/libCEED/blob/master/CONTRIBUTING.md) file.), in alphabetical order by last name. However, the CU Boulder's libCEED team (cc: @LeilaGhaffari, @jeremylt ) really thinks that @jedbrown should be the first author for this first libCEED's whitepaper. So we can keep the chosen convention, but with the exception that Jed is the first author, if everybody is fine with that.

Reviewer: With Jed being the architect for the library and the long term maintainer, it makes sense to me for him to be listed as first author on the first libCEED paper.

Reviewer 5: Yes, I agree. It seems fair to me.

Reviewer 6: I think that's a good plan.

  - name: David Medina
    affiliation: 4
  - name: Will Pazner
    affiliation: 2
    orcid: 0000-0003-4885-2934

>   - name: Thilina Rathnayake

Author: Wow! We had a couple of names misspelled! I'm sorry! Considering that I had taken them from the technical reports... ^_^''

Reviewer 2: Not a mistake per se; he changed the way he anglicized the name and we've been slow to update consistently.

   index: 2
 - name: University of Tennessee
   index: 3

>  - name: Occalytics LLC

Author: @dmed256 I had found a different affiliation for you on the internet, but Tzanio suggested Occalytics LLC for the MS-34 report. Could you please confirm or change that at your convenience? Thanks

The Julia interface, referred to as ``LibCEED.jl``, provides both a low-level interface, which is generated automatically from ``libCEED``'s C header files, and a high-level interface. The high-level interface takes advantage of Julia features such as garbage collection to simplify memory management. A key feature of ``LibCEED.jl`` is the straightforward creation of user QFunctions that work on both the CPU and GPU, making use of Julia's metaprogramming and just-in-time compilation capabilities [@Julia-2017].

The Rust interface provides a high-level interface which wraps the automatically generated bindings from the ``libCEED`` C header files. The Rust interface takes advantage of ownership and borrow checking in Rust to provide improved memory safety and uses closures to simplify management of user provided environment variables in the context of the user QFunctions.

> 

Author: Thanks @jeremylt ! Is there a Rust reference to cite here?

Reviewer: Not that I'm aware of? We aren't using anything that isn't a core language feature.

Author: I meant even just a dev website, if there is no paper or no official way to cite them explicitly mentioned on their website (like we did for other languages or packages we mentioned in this paper). We should not assume every reader is familiar with what Rust is.

Reviewer: Do we want to also include citations on what Python, C, and Fortran77 are? Those are the other languages mentioned but not cited directly.

Author: Ahah that is true. I just thought it might be a bit unheard of for some readers since it's relative new. I am always in favor of more references than less :)

Reviewer: I unified our approach to citing languages. What do you think?

Author: Thanks Jeremy. I thought you were being sarcastic when you mentioned to cite all of the languages 😂 ah, the joy of online conversations with no tonal and/or visual aid. 

To me it's ok. I guess we can hear @jedbrown 's opinion, too, when he gets the chance.

Reviewer: I was sort of on the fence, but you like references and I like uniformity, so it seemed a good way to satisfy both.

Reviewer 2: Seems fine to me.
